subsys/mgmt/mcumgr/smp_pkasm.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_MGMT_SMP_PKASM_H_
+#define ZEPHYR_INCLUDE_MGMT_SMP_PKASM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct zephyr_smp_transport;
+
+/**
+ * Collect data to re-assembly buffer
+ *
+ * The function adds data to the end of current re-assembly buffer; it will allocate new buffer
+ * if there isn't one allocated.
+ *
+ * Note: Currently the function is not able to concatenate buffers so re-assembled packet needs
+ * to fit into one buffer.
+ *
+ * @param zst	the SMP transport;
+ * @param buf	buffer with data to add;
+ * @param len	length of data to add;
+ *
+ * @return	number of expected bytes left to complete the packet, 0 means buffer is complete
+ *		and no more fragments are expected;
+ *		-ENOSR if packet length read from header is bigger than CONFIG_MCUMGR_BUF_SIZE,
+ *		which means there is not way to fit it in configured buffer;
+ *		-EOVERFLOW if attempting to add fragment that would make complete packet larger
+ *		than expected, by length read from a header;
+ *		-ENOMEM if failed to allocate new buffer for packet assembly;
+ *		-ENODATA if first received fragment was not big enough to figure out size of packet,
+ *		MTU is set too low;
+ */
+int zephyr_smp_pkasm_collect(struct zephyr_smp_transport *zst, const void *buf, uint16_t len);
+
+/**
+ * Return number of expected bytes to complete the packet
+ *
+ * @param zst	the SMP transport;
+ *
+ * @return	number of bytes needed to complete the packet;
+ *		-EINVAL if there is no packet in re-assembly;
+ */
+static inline int zephyr_smp_pkasm_expected(const struct zephyr_smp_transport *zst)
+{
+	if (zst->_pkasm.current == NULL) {
+		return -EINVAL;
+	}
+
+	return zst->_pkasm.expected;
+}
+
+/**
+ * Pass packet for further processing
+ *
+ * Checks if packet has enough data to be deemed re-assembled and passes it for further processing.
+ * If successful then the re-assembly state in @p zst will indicate that there is no
+ * re-assembly in progress.
+ * The function can be forced to pass data for processing even if packet is not complete,
+ * in which case it is users responsibility to use user data, passed with packet, to notify
+ * receiving end of such case.
+ *
+ * @param zst	the SMP transport;
+ * @param force	process anyway;
+ *
+ * @return	0 on success;
+ *		-EINVAL if there is not re-assembly in progress;
+ *		-ENODATA if there is not enough data to consider packet re-assembled, it has not
+ *		been passed further.
+ */
+int zephyr_smp_pkasm_complete(struct zephyr_smp_transport *zst, bool force);
```

Comment:
It seems that this API function is unnecessary and its functionality can be incorporated into zephyr_smp_pkasm_collect. This way, we will simplify API for this module and get rid of the EINVAL opcode which can happen due to API misuse.

Reply:
This is intentionally so that the `collect` caller can decide whether it want's to add something to the "user data" or maybe not pass the frame for completion, for some reason.
Same reason for "force" so that transport could decide if broken frame should be passed up or not.